plugin/evm/message/handshake/upgrade_config_message.go

```diff
@@ -36,45 +24,32 @@ func (u *UpgradeConfigMessage) ID() common.Hash {
 	return u.hash
 }
 
-// Attempts to parse a networkUpgradeConfigMessage from a []byte
-//
-// This function attempts to parse a stream of bytes as a
-// networkUpgradeConfigMessage (as serialized from
-// UpgradeConfigToNetworkMessage).
+// Attempts to parse a params.UpgradeConfig from a []byte
 //
 // The function returns a reference of *params.UpgradeConfig
-func NewUpgradeConfigMessageFromBytes(bytes []byte) (*params.UpgradeConfig, error) {
-	var config networkUpgradeConfigMessage
-	version, err := Codec.Unmarshal(bytes, &config)
+func UpgradeConfigFromBytes(bytes []byte) (*params.UpgradeConfig, error) {
+	var upgradeConfig params.UpgradeConfig
+	err := json.Unmarshal(bytes, &upgradeConfig)
 	if err != nil {
 		return nil, err
 	}
-	if version != Version {
-		return nil, ErrInvalidVersion
-	}
 
-	var PrecompileUpgrades []params.PrecompileUpgrade
-	for _, precompileUpgrade := range config.PrecompileUpgrades {
-		module, ok := modules.GetPrecompileModule(precompileUpgrade.Key)
-		if !ok {
-			return nil, ErrUnknowPrecompile
-		}
-		preCompile := module.MakeConfig()
-		version, err := Codec.Unmarshal(precompileUpgrade.Bytes, preCompile)
-		if version != Version {
-			return nil, ErrInvalidVersion
-		}
-		if err != nil {
-			return nil, err
-		}
-		PrecompileUpgrades = append(PrecompileUpgrades, params.PrecompileUpgrade{Config: preCompile})
-	}
+	return &upgradeConfig, nil
+}
 
-	return &params.UpgradeConfig{
-		OptionalNetworkUpgrades: config.OptionalNetworkUpgrades,
-		StateUpgrades:           config.StateUpgrades,
-		PrecompileUpgrades:      PrecompileUpgrades,
-	}, nil
+// Encodes any object to JSON with a deterministic output. All the keys, even
+// into inner objects, are sorted alphabetically.
```

Comment:
Please link docs/code from the stdlib that explain why this is correct. I found this documentation in `encoding/json`, for example:
```
Map values encode as JSON objects. The map's key type must either be a string, an integer type, or implement encoding.TextMarshaler. The map keys are sorted and used as JSON object keys by applying the following rules, subject to the UTF-8 coercion described for string values above:
```

Reply:
The most important line seems to be this one (inside the `mapEncoder`: https://cs.opensource.google/go/go/+/refs/tags/go1.21.5:src/encoding/json/encode.go;l=751)

Reply:
All in all, this is an interesting property to rely on as it goes above what is required in the JSON spec: https://www.ietf.org/rfc/rfc4627.txt

Reply:
its pretty crazy they do this by default instead of putting it behind a config flag (because there is an obvious performance penalty)

Reply:
(I don't think your implementation is wrong to be clear, just potentially brittle)